worker/changestream/eventqueue/eventqueue.go

```diff
@@ -144,63 +139,34 @@ func (q *EventQueue) unsubscribe(subscriptionID uint64) {
 	}
 
 	delete(q.subscriptions, subscriptionID)
-
-	// Close the subscription channel after a dispatch. If a unsubscription
-	// happens during a dispatch, then it causes the event queue to potentially
-	// panic. Removal of the subscription from the queue is synchronous, so
-	// the subscription can no longer be used when another dispatch occurs. The
-	// closing of the channel then becomes asynchronous and can be closed during
-	// the next selection.
-	q.tomb.Go(func() error {
-		// Ensure that we give up if the tomb is dying.
-		select {
-		case <-q.tomb.Dying():
-			return tomb.ErrDying
-		case q.actions <- func() {
-			close(sub.changes)
-		}:
-		}
-		return nil
-	})
 }
 
 func (q *EventQueue) loop() error {
-	defer func() {
-		q.mutex.Lock()
-		defer q.mutex.Unlock()
-
-		for _, sub := range q.subscriptions {
-			close(sub.changes)
-		}
-		q.subscriptions = make(map[uint64]*subscription)
-	}()
-
 	for {
-		var (
-			ch changestream.ChangeEvent
-			ok bool
-		)
 		select {
 		case <-q.tomb.Dying():
 			return tomb.ErrDying
-		case action := <-q.actions:
-			action()
-			continue
-		case ch, ok = <-q.stream.Changes():
+		case ch, ok := <-q.stream.Changes():
 			// If the stream is closed, we expect that a new worker will come
 			// again using the change stream worker infrastructure. In this case
 			// just ignore and close out.
 			if !ok {
 				q.logger.Infof("change stream change channel is closed")
 				return nil
 			}
-		}
 
-		// The dispatch is done in two stages to avoid holding the lock for the
-		// duration of the dispatch.
-		subs := q.gatherSubscriptions(ch)
-		if err := q.dispatchChange(subs, ch); err != nil {
-			return errors.Trace(err)
+			subs := q.gatherSubscriptions(ch)
+			for _, sub := range subs {
+				// Ensure we check tomb dying as the handling logic is
+				// synchronous and blocking. This is to effectively handle
+				// back pressure on the stream.
+				select {
+				case <-q.tomb.Dying():
+					return tomb.ErrDying
+				default:
+				}
+				sub.handler(ch)
```

Comment:
Only problem I see here is that even if Unsubscribe is called (and returned), the handler can still be called at most once more. I think its fine for Unsubscribe to block until the subscription is no longer used.

Something like

```
q.handlerWaitMut.Lock()
subs := q.gatherSubscriptions(ch)
for _, sub := range subs {
	// Ensure we check tomb dying as the handling logic is
	// synchronous and blocking. This is to effectively handle
	// back pressure on the stream.
	select {
	case <-q.tomb.Dying():
		q.handlerWaitMut.Unlock()
		return tomb.ErrDying
	default:
	}
	sub.handler(ch)
}
q.handlerWaitMut.Unlock()
```

```
func (q *EventQueue) unsubscribe(subscriptionID uint64) {
	q.mutex.Lock()
	sub, found := q.subscriptions[subscriptionID]
	if !found {
		q.mutex.Unlock()
		return
	}
	for topic := range sub.topics {
		var updatedFilters []*eventFilter
		for _, filter := range q.subscriptionsByNS[topic] {
			if filter.subscriptionID == subscriptionID {
				continue
			}
			updatedFilters = append(updatedFilters, filter)
		}
		q.subscriptionsByNS[topic] = updatedFilters
	}

	delete(q.subscriptions, subscriptionID)
	q.mutex.Unlock()
	q.handlerWaitMut.RLock()
	q.handlerWaitMut.RUnlock()
```

Reply:
I need to think about this more. I'm currently on the fence that an unsubscribe will only take effect after the current tick. I can see why you'd want to potentially block a dispatch though. I'm not a fan of an unsubscribe being used a way to prevent a dispatch at the last second. I've got a suspicion that Hyrum's law will come into play.

Reply:
Anyway, Unsubscribe could be effectively synchronous now (minus the go routine in unsubscribe). Which guarantees the callback won't be called after the Unsubscribe method returns. Which means I am kind of happy with this.